### ISSUE TITLE
[Native Image] Minecraft server is successfully built to native executable, but it doesn't accept/receive any incoming connections

### ISSUE THREAD

**Issue by BlueGradientHorizon** [oracle/graal#10224]
### Describe the Issue

Hello. I've been trying to build native image of most recent version of Minecraft server fork [PaperMC](https://github.com/PaperMC/Paper) and achieved some sort of success. The server is running, generating all of needed files, chunks, executing commands, etc, but it doesn't accept any clients. Nothing is being printed in the console when client tries to connect.

### Using the latest version of GraalVM can resolve many issues.

- [X] I tried with the [latest version](https://www.graalvm.org/downloads/) of GraalVM.

### GraalVM Version

```
java version "24" 2025-03-18
Java(TM) SE Runtime Environment Oracle GraalVM 24-dev+23.1 (build 24+23-jvmci-b01)
Java HotSpot(TM) 64-Bit Server VM Oracle GraalVM 24-dev+23.1 (build 24+23-jvmci-b01, mixed mode, sharing)
```
[Oracle GraalVM for JDK 24 - early access build - 24.0.0-ea.22](https://github.com/graalvm/oracle-graalvm-ea-builds/releases/tag/jdk-24.0.0-ea.22)

### Operating System and Version

Windows 11 23H2 22631.4317 amd64

### Troubleshooting Confirmation

- [X] I tried the suggestions in the [troubleshooting guide](https://www.graalvm.org/latest/reference-manual/native-image/guides/troubleshoot-run-time-errors/).

### Run Command

See `Steps to Reproduce` section.

### Expected Behavior

To be able for clients to connect to the native-image-compiled PaperMC server.

### Actual Behavior

Clients can't connect to the server, including getting its MOTD, like it isn't even running.

### Steps to Reproduce

0. I'm on Windows, and JFR isn't supported ([source](https://www.graalvm.org/latest/reference-manual/native-image/debugging-and-diagnostics/JFR/)):

> Note: JFR event recording is not yet available with Native Image on Windows.

But, interestingly, although I didn't enable it via passing cmd arguments to native-image, mc server was crashing.
<details>
<summary>Stacktrace</summary>

```
[18:27:55] [main/WARN]: Failed to load datapacks, can't proceed with server load. You can either fix your datapacks or reset to vanilla with --safeMode
java.util.concurrent.ExecutionException: java.lang.UnsatisfiedLinkError: jdk.jfr.internal.JVM.isExcluded(Ljava/lang/Class;)Z [symbol: Java_jdk_jfr_internal_JVM_isExcluded or Java_jdk_jfr_internal_JVM_isExcluded__Ljava_lang_Class_2]
	at java.base/java.util.concurrent.CompletableFuture.wrapInExecutionException(CompletableFuture.java:347) ~[org.bukkit.craftbukkit.main.exe:?]
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:442) ~[org.bukkit.craftbukkit.main.exe:?]
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2119) ~[org.bukkit.craftbukkit.main.exe:?]
	at net.minecraft.server.Main.main(Main.java:304) ~[?:?]
	at io.papermc.paper.PaperBootstrap.boot(PaperBootstrap.java:21) ~[?:?]
	at org.bukkit.craftbukkit.Main.main(Main.java:281) ~[?:?]
	at java.base/java.lang.invoke.LambdaForm$DMH/sa346b79c.invokeStaticInit(LambdaForm$DMH) ~[?:?]
Caused by: java.lang.UnsatisfiedLinkError: jdk.jfr.internal.JVM.isExcluded(Ljava/lang/Class;)Z [symbol: Java_jdk_jfr_internal_JVM_isExcluded or Java_jdk_jfr_internal_JVM_isExcluded__Ljava_lang_Class_2]
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.jni.access.JNINativeLinkage.getOrFindEntryPoint(JNINativeLinkage.java:152) ~[?:?]
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.jni.JNIGeneratedMethodSupport.nativeCallAddress(JNIGeneratedMethodSupport.java:41) ~[?:?]
	at jdk.jfr@24/jdk.jfr.internal.JVM.isExcluded(Native Method) ~[?:?]
	at jdk.jfr@24/jdk.jfr.internal.MetadataRepository.register(MetadataRepository.java:139) ~[?:?]
	at jdk.jfr@24/jdk.jfr.internal.MetadataRepository.register(MetadataRepository.java:134) ~[?:?]
	at jdk.jfr@24/jdk.jfr.FlightRecorder.register(FlightRecorder.java:129) ~[org.bukkit.craftbukkit.main.exe:?]
	at java.base/java.lang.Iterable.forEach(Iterable.java:75) ~[org.bukkit.craftbukkit.main.exe:?]
	at net.minecraft.util.profiling.jfr.JfrProfiler.<init>(JfrProfiler.java:79) ~[?:?]
	at net.minecraft.util.profiling.jfr.JfrProfiler.<clinit>(JfrProfiler.java:79) ~[?:?]
	at net.minecraft.util.profiling.jfr.JvmProfiler.<clinit>(JvmProfiler.java:49) ~[org.bukkit.craftbukkit.main.exe:1.21.3-80-c2294d7]
	at net.minecraft.commands.Commands.<init>(Commands.java:225) ~[org.bukkit.craftbukkit.main.exe:1.21.3-80-c2294d7]
	at net.minecraft.server.ReloadableServerResources.<init>(ReloadableServerResources.java:41) ~[?:?]
	at net.minecraft.server.ReloadableServerResources.lambda$loadResources$1(ReloadableServerResources.java:85) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1196) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:556) ~[org.bukkit.craftbukkit.main.exe:?]
	at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:660) ~[org.bukkit.craftbukkit.main.exe:?]
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:699) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:528) ~[org.bukkit.craftbukkit.main.exe:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[org.bukkit.craftbukkit.main.exe:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.base/java.lang.Thread.runWith(Thread.java:1589) ~[org.bukkit.craftbukkit.main.exe:?]
	at java.base/java.lang.Thread.run(Thread.java:1576) ~[org.bukkit.craftbukkit.main.exe:?]
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:832) [org.bukkit.craftbukkit.main.exe:?]
	at org.graalvm.nativeimage.builder/com.oracle.svm.core.thread.PlatformThreads.threadStartRoutine(PlatformThreads.java:808) [org.bukkit.craftbukkit.main.exe:?]
```
</details>

Later I've checked how native executable will behave if compiled for linux aarch64 (Termux, ubuntu proot-distro) using the same GraalVM EA version, but I got the same exception message. Anyway I solved this problem by creating a [patch](https://github.com/BlueGradientHorizon/mc-native-image/blob/main/1073-Disable-JFR-for-Native-Image-support.patch) for PaperMC and [compiling it from source](https://github.com/PaperMC/Paper/?tab=readme-ov-file#how-to-compiling-jar-from-source).

1. Run the server bootstrapper for the first time to unpack libraries and patch official server with Paper patches
```
"C:\Program Files\graalvm-jdk-24+23.1\bin\java.exe" -jar paper-paperclip-1.21.3-R0.1-SNAPSHOT-mojmap.jar
```

2. Generate the list of needed libraries for classpath argument
```powershell
(ls .\libraries\ -Recurse -Include *.jar | % { $_.FullName }) -Split([Environment]::NewLine, [System.StringSplitOptions]::RemoveEmptyEntries) -join ";" > libraries.txt
```

3. Create file `arg.txt` with classpath and main launch class for convenience
```
-cp versions\1.21.3\paper-1.21.3.jar;<contents of libraries.txt> org.bukkit.craftbukkit.Main
```

5. Run the server with metadata collection
```
"C:\Program Files\graalvm-jdk-24+23.1\bin\java.exe" -agentlib:native-image-agent=config-merge-dir=./,config-write-period-secs=10 @arg.txt nogui
```
First, you'll need to accept [Minecraft EULA](https://aka.ms/MinecraftEULA). Edit `eula.txt` and set `eula` value to `true` and then run the server again using the same command.

6. Slightly modify `reachability-metadata.json`
By some reason, native image agent does not include all files of the data directory inside server jar into that file, but these are crucial for server to run, so replace:
```json
    {
      "glob": "data/.mcassetsroot"
    },
```
with
```json
    {
      "glob": "data/**"
    },
```

7. Build native image
```
"C:\Program Files\graalvm-jdk-24+23.1\bin\native-image.cmd" --no-fallback --enable-url-protocols=http,https --initialize-at-build-time=net.minecraft.util.profiling.jfr.event,org.apache.logging.log4j,net.minecrell.terminalconsole.TCALookup -H:+UnlockExperimentalVMOptions -H:ConfigurationFileDirectories=./ -H:+AddAllCharsets -H:+IncludeAllLocales @arg.txt
```

8. Run native executable
```
org.bukkit.craftbukkit.main.exe nogui
```
`nogui` is mandatory or you'll get `java.lang.NoSuchMethodError: java.awt.Toolkit.getDefaultToolkit()Ljava/awt/Toolkit;` (by default it will try to open GUI window with console and memory monitor).

9. Try connecting to the server using official unmodified Minecraft 1.21.3 client
You'll not be able to do so, like the server isn't even running. But it actually opens a default 25565 port (I've checked it by using [TCPView](https://learn.microsoft.com/en-us/sysinternals/downloads/tcpview)).

### Additional Context

PaperMC server (commit [`da71382`](https://github.com/PaperMC/Paper/commit/da7138233f6392e791d790d1c3407414c855f9c2)) with my patch - [link](https://drive.google.com/file/d/1rgrPxFdZEkX7UrtnX7D27Dt3oJgYcbeh/view?usp=sharing)
Compiled native image of that server for Windows amd64 - [link](https://drive.google.com/file/d/1RZfjJWwvxBwjruXUzSvFPWhI9JNncEuD/view?usp=sharing)
Native image bundle - [link](https://drive.google.com/file/d/19KNxkNfH9bw6DUw4iuSOAoACiZGAmmQ1/view?usp=sharing)

Minecraft uses Netty for network communications.

### Run-Time Log Output and Error Messages

Nothing is being printed in the console when client tries to connect.

**Comment by selhagani** [comment]
Hi @BlueGradientHorizon,

Thank you for reaching out!

Since there’s no specific error log provided, it’s difficult to pinpoint exactly what went wrong with the native executable. It’s possible that despite using the tracing agent, there are still missing metadata registration issues. I recommend diagnosing this by following the steps outlined in [this documentation](https://www.graalvm.org/latest/reference-manual/native-image/guides/troubleshoot-run-time-errors/). These steps can help identify and resolve missing metadata problems.

Unfortunately, I’m unable to reproduce this on my end as I don’t have a Minecraft account.

**Comment by thejudge156** [comment]
Hey! I work on MCNative, a (albeit work-in-progress) project that does the MC Client and (soon) different server software. I have some Proof-of-concept in my repo, but you should look at these [files](https://github.com/thejudge156/MCNativeBuilder/tree/refactor/mc-lib/src/main/resources/META-INF/native-image/me.judge/mc-lib). I suspect the issue is not these files that I linked however. If I remember correctly, you have to run the metadata agent and *join* the server on a *client* and then stop the server. This is also not good enough for stuff like the Ender Dragon, (which uses reflection for some reason). So if you want to even go to the end, you should run the following command:
`/execute in minecraft:the_end run tp @s 0 128 0`

Just make sure you are in creative, then let the ender dragon do its thing for like 3 minutes. Then you can stop the server and fight it in the native image.

If you would like to join forces on this, let me know. I am more than happy to share my knowledge and experience.

**Comment by thejudge156** [comment]
Also, rather than manually patching the jar, take a look at this [substitution ](https://github.com/thejudge156/MCNativeBuilder/blob/refactor/mc-lib/src/main/java/me/judge/mclib/JFRSubstitutions.java) I made using GraalVM's sub api.

**Comment by BlueGradientHorizon** [comment]
I solved the problem. I want to thank @thejudge156, because he pointed me right. In order to be able to build a more correct native executable, it was needed more metadata. After I launched the server with tracing agent I connected to it via Minecraft client and did some actions: primarily chunk generation and joining into other worlds. `reachability_metadata.json` size has grown from ~160kb to ~190kb. Then I've rebuilt native image and everything started working correctly. Now I was able to get server's MOTD, connect to it, generate new chunks, join other worlds - generally, perform the same actions I've done before on Java server with tracing agent.
Actually, I already was doing the exact thing when I only started this experiment, but since I ran into the problem about JFR I've completely forgotten what I needed to generate metadata again (I was cleaning the working directory several times during testing, including this file).
But now I've experiencing another problem: when building native image with `--pgo-instrument` option, server crashes at runtime, after line `Initialized 0 plugins` with error message: `WaitForSingleObject failed with error 0x57`. I'm closing this issue and opening a new one.
